soroban-env-host/tests/ed25519_edge_cases.rs

```diff
@@ -0,0 +1,1015 @@
+//!     "canonical" encoding involves reducing the y-coordinate modulo 2^255-19,
+//!     the "non-canonical" encoding does not. There are 19 possible
+//!     non-canonical point encodings and of which 2 have small order, 10 have
+//!     mixed order (i.e. have nonzero torsion-subgroup components _and_ nonzero
+//!     prime-subgroup components), and 7 decode to invalid curve points.
+//!
+//!   - It is also possible to encode the scalar used in the signature equation
+//!     in different ways -- again reducing modulo the group order L, or not --
+//!     but in practice all libraries reject the non-reduced ("non-canonical")
+//!     scalar encodings.
+//!
+//! If you want to read a treatment of these issues in more detail or by people
+//! who are better at math, read the blog post
+//! https://hdevalence.ca/blog/2020-10-04-its-25519am as well as the paper
+//! https://eprint.iacr.org/2020/1244.pdf
+//!
+//! The Zcash project noticed the remaining variation, and published ZIP-215 to
+//! specify what to do. They chose to _accept_ all the non-canonical encodings
+//! and mandate the cofactor in the verification equation (i.e. multiplying by
+//! 8). They made these choices for the sake of (a) supporting batch
+//! verification and (b) supporting existing signatures they have in the field
+//! under existing code, some of which might have nonzero torsion-subgroup
+//! components or non-canonical encodings.
+//!
+//! Since Soroban is a new implementation based on the dalek libraries, and
+//! because we wish to also interoperate with stellar-core which is based on the
+//! libsodium library, we actually take the opposite position from ZIP-215:
+//!
+//!   - reject all small-order points (though accept mixed-order)
+//!   - reject all non-canonical encodings
+//!   - verify without the cofactor in the verification equation
+//!
+//! This is the default behaviour in libsodium (at least after version 1.0.16-18
+//! which changed the rules, which we use in stellar-core) as well as dalek's
+//! `verify_strict` function. We test dalek in this file and have a
+//! corresponding set of tests with the same test vectors in stellar-core to
+//! confirm identical behaviour with libsodium.
+
+use ed25519_dalek::{Signature, VerifyingKey, PUBLIC_KEY_LENGTH, SIGNATURE_LENGTH};
+
+// The Zcash work produced 14 x 14 = 196 test vectors expressing combinations
+// of boundary cases for public keys and signatures (14 here being the 8
+// canonical encodings of the 8 small-order points, plus 6 non-canonical
+// encodings of some of these points; and then there's a test vector for such a
+// small-order point occurring in either the public key _or_ the signature, so
+// 14 x 14). We use these test vectors to verify that our implementation
+// _rejects_ them all.
+#[test]
+fn check_zcash_test_vectors_are_rejected() {
+    for (i, test_vector) in ZCASH_TEST_VECTORS.iter().enumerate() {
+        println!("IACR 2020/1244 Test vector {}", i);
```

Comment:
This log line references IACR but this test is testing Zcash test vectors.

Reply:
Oops, fixed